arangod/Replication2/ReplicatedState/Supervision.cpp

```diff
@@ -21,27 +21,139 @@
+      << "expected = " << expectedGeneration
+      << " planned = " << plan.participants.at(participant).generation;
+  if (auto iter = current.participants.find(participant);
+      iter != current.participants.end()) {
+    auto const& state = iter->second;
+    return state.generation == expectedGeneration &&
+           state.snapshot.status == SnapshotStatus::kCompleted;
+  }
+
+  return false;
+}
+
+auto isParticipantSnapshotCompleted(ParticipantId const& participant,
+                                    RSA::Current const& current,
+                                    RSA::Plan const& plan) -> bool {
+  if (auto iter = plan.participants.find(participant);
+      iter != plan.participants.end()) {
+    auto expectedGeneration = iter->second.generation;
+    return isParticipantSnapshotCompleted(participant, expectedGeneration,
+                                          current, plan);
+  }
+
+  return false;
+}
+
+/**
+ * A server is considered OK if
+ * - its snapshot is complete
+ * - and is allowedAsLeader && allowedInQuorum in Log/Target and Log/Plan
+ * @param participant
+ * @param log
+ * @param state
+ * @return
+ */
+auto isParticipantOk(ParticipantId const& participant, RLA::Log const& log,
+                     RSA::State const& state) {
+  TRI_ASSERT(state.current.has_value());
+  TRI_ASSERT(state.plan.has_value());
+  TRI_ASSERT(log.plan.has_value());
+
+  // check if the participant has an up-to-date snapshot
+  auto snapshotOk =
+      isParticipantSnapshotCompleted(participant, *state.current, *state.plan);
+  if (!snapshotOk) {
+    return false;
+  }
+
+  auto const flagsAreCorrect = [&](ParticipantsFlagsMap const& flagsMap) {
+    if (auto iter = flagsMap.find(participant); iter != flagsMap.end()) {
+      auto const& flags = iter->second;
+      return flags.allowedAsLeader and flags.allowedInQuorum;
+    }
+    return true;
+  };
+
+  // check if the flags for that participant are set correctly
+  auto const& config = log.plan->participantsConfig.participants;
+  return flagsAreCorrect(config) and flagsAreCorrect(log.target.participants);
+}
+
+/**
+ * Counts the number of OK participants.
+ * @param log
+ * @param state
+ * @return
+ */
+auto countOkServers(RLA::Log const& log, RSA::State const& state)
+    -> std::size_t {
+  return std::count_if(
+      state.plan->participants.begin(), state.plan->participants.end(),
+      [&](auto const& p) { return isParticipantOk(p.first, log, state); });
+}
+
+auto checkStateAdded(SupervisionContext& ctx, RSA::State const& state) {
   auto id = state.target.id;
 
   if (!state.plan) {
-    auto statePlan = replication2::replicated_state::agency::Plan{
-        .id = state.target.id,
-        .generation = StateGeneration{1},
-        .properties = state.target.properties,
-        .participants = {}};
+    auto statePlan = RSA::Plan{.id = state.target.id,
+                               .generation = StateGeneration{2},
```

Comment:
Because generation 1 is used by the initial participants. And the algorithm is, at least at the moment, to use the number written there for the new participant and then increment it. Having 1 here would add a new participant with generation 1.

Reply:
Maybe a comment clarifying this wouldn't go amiss ;)